mecanum_drive_controller/src/mecanum_drive_controller.cpp

```diff
@@ -0,0 +1,506 @@
+}
+
+controller_interface::return_type MecanumDriveController::update_reference_from_subscribers(
+  const rclcpp::Time & time, const rclcpp::Duration & /*period*/)
+{
+  auto current_ref = *(input_ref_.readFromRT());
+  const auto age_of_last_command = time - (current_ref)->header.stamp;
+
+  // send message only if there is no timeout
+  if (age_of_last_command <= ref_timeout_ || ref_timeout_ == rclcpp::Duration::from_seconds(0))
+  {
+    if (
+      !std::isnan(current_ref->twist.linear.x) && !std::isnan(current_ref->twist.linear.y) &&
+      !std::isnan(current_ref->twist.angular.z))
+    {
+      reference_interfaces_[0] = current_ref->twist.linear.x;
+      reference_interfaces_[1] = current_ref->twist.linear.y;
+      reference_interfaces_[2] = current_ref->twist.angular.z;
+
+      if (ref_timeout_ == rclcpp::Duration::from_seconds(0))
+      {
+        current_ref->twist.linear.x = std::numeric_limits<double>::quiet_NaN();
+        current_ref->twist.linear.y = std::numeric_limits<double>::quiet_NaN();
+        current_ref->twist.angular.z = std::numeric_limits<double>::quiet_NaN();
+      }
+    }
+  }
+  else
+  {
+    if (
+      !std::isnan(current_ref->twist.linear.x) && !std::isnan(current_ref->twist.linear.y) &&
+      !std::isnan(current_ref->twist.angular.z))
+    {
+      reference_interfaces_[0] = 0.0;
+      reference_interfaces_[1] = 0.0;
+      reference_interfaces_[2] = 0.0;
+
+      current_ref->twist.linear.x = std::numeric_limits<double>::quiet_NaN();
+      current_ref->twist.linear.y = std::numeric_limits<double>::quiet_NaN();
+      current_ref->twist.angular.z = std::numeric_limits<double>::quiet_NaN();
+    }
+  }
+
+  return controller_interface::return_type::OK;
+}
+
+controller_interface::return_type MecanumDriveController::update_and_write_commands(
+  const rclcpp::Time & time, const rclcpp::Duration & period)
+{
+  // FORWARD KINEMATICS (odometry).
+  double wheel_front_left_vel = state_interfaces_[0].get_value();
+  double wheel_back_left_vel = state_interfaces_[1].get_value();
+  double wheel_back_right_vel = state_interfaces_[2].get_value();
+  double wheel_front_right_vel = state_interfaces_[3].get_value();
+
+  if (
+    !std::isnan(wheel_front_left_vel) && !std::isnan(wheel_back_left_vel) &&
+    !std::isnan(wheel_back_right_vel) && !std::isnan(wheel_front_right_vel))
+  {
+    // Estimate twist (using joint information) and integrate
+    odometry_.update(
+      wheel_front_left_vel, wheel_back_left_vel, wheel_back_right_vel, wheel_front_right_vel,
+      period.seconds());
+  }
+
+  // INVERSE KINEMATICS (move robot).
+  // Compute wheels velocities (this is the actual ik):
+  // NOTE: the input desired twist (from topic `~/reference`) is a body twist.
+  auto current_ref = *(input_ref_.readFromRT());
+  const auto age_of_last_command = time - current_ref->header.stamp;
+  if (age_of_last_command <= ref_timeout_ || ref_timeout_ == rclcpp::Duration::from_seconds(0))
+  {
```

Comment:
Why is this here again! This breaks behavior of chained controllers…



```suggestion
  if (
    !std::isnan(reference_interfaces_[0]) && !std::isnan(reference_interfaces_[1]) &&
    !std::isnan(reference_interfaces_[2]))
  {
```